### PR TITLE
CTCTRADERS-729: Set MesSenMES3 to X-Message-Sender (Backend only)

### PR DESCRIPTION
**Pull request by kenneth-rayner**
Work to set MesSenMES3 in backend

- Filter used to return bad request if frontend/api includes post with MesSenMES3 included
- MesSenMES3 is added to the message before saving/sending on to EIS
- Future refactor work to put the transformation into a custom action (tried this but some controllers don't have an arrival id available)

### CODE REVIEW

File: test/controllers/MessagesControllerSpec.scala [hunk 1]
```diff
@@ -139,8 +142,14 @@ class MessagesControllerSpec extends SpecBase with ScalaCheckPropertyChecks with
...
                                                                    eqTo(MessageId.fromIndex(1)),
-                                                                   eqTo(movementMessge),
+                                                                   captor.capture(),
                                                                    eqTo(ArrivalStatus.UnloadingRemarksSubmitted))(any())
+          val movement: MovementMessageWithStatus = captor.getValue
+          movement.messageCorrelationId mustEqual movementMessage.messageCorrelationId
+          movement.status mustEqual movementMessage.status
+          movement.messageType mustEqual movementMessage.messageType
+          movement.dateTime mustEqual movementMessage.dateTime
+          movement.message.map(trim) mustEqual movementMessage.message.map(trim)
```
[the remark this answers is not in the document]
Comment (chris-ianson) [reply test/controllers/MessagesControllerSpec.scala]: Updated to check entire movement object

File: app/controllers/MovementsController.scala [hunk 2]
```diff
@@ -65,12 +64,13 @@ class MovementsController @Inject()(
...
 
-  def post: Action[NodeSeq] = authenticatedOptionalArrival().async(parse.xml) {
+  def post: Action[NodeSeq] = (authenticatedOptionalArrival()(parse.xml) andThen validateMessageSenderNode.filter).async {
     implicit request =>
       request.arrival match {
         case Some(arrival) if allMessageUnsent(arrival.messages) =>
+          val updatedXml = XMLTransformer.updateMesSenMES3(arrival.arrivalId, arrival.nextMessageCorrelationId, request.body)
           arrivalMovementService
-            .makeMovementMessageWithStatus(arrival.nextMessageCorrelationId, MessageType.ArrivalNotification)(request.body)
+            .makeMovementMessageWithStatus(arrival.nextMessageCorrelationId, MessageType.ArrivalNotification)(updatedXml)
```
[the remark this answers is not in the document]
Comment (chris-ianson) [reply app/controllers/MovementsController.scala]: Moved the functionality into ArrivalMovementMessageService. Signature on makeArrivalMovement has changed (not sure if you want to refactor the others now)

File: app/utils/XMLTransformer.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  def addXmlNode(existingNode: String, key: String, value: String, inputXml: NodeSeq): NodeSeq =
+    createRuleTransformer(existingNode, key, value).transform(inputXml.head)
+
+  def updateMesSenMES3(arrivalId: ArrivalId, correlationId: Int, body: NodeSeq): NodeSeq =
+    if ((body \\ "SynVerNumMES2").nonEmpty) {
+      val messageSender: MessageSender = MessageSender(arrivalId, correlationId)
+      XMLTransformer.addXmlNode("SynVerNumMES2", "MesSenMES3", messageSender.toString, body)
+    } else {
+      Logger.warn("Couldn't find SynVerNumMES2 node")
```
[the remark this answers is not in the document]
Comment (chris-ianson) [reply app/utils/XMLTransformer.scala]: Refactored XMLTransformer. Only a single method is needed to add the new node - updateMesSenMES3 (I'm starting to think updateMesSenMES3 should be its own class that extends off XMLTransformer but that's probably for another branch)

File: app/services/ArrivalMovementMessageService.scala [hunk 4]
```diff
@@ -22,39 +22,35 @@ import com.google.inject.Inject
...
+  import XMLTransformer._
   import XmlMessageParser._
 
-  private[this] def nodeSeqToEither(xml: NodeSeq): ParseHandler[NodeSeq] =
-    if (xml != null) {
-      Right(xml)
-    } else {
-      Left(EmptyNodeSeq("Request body is empty"))
-    }
+  def makeArrivalMovement(eori: String, nodeSeq: NodeSeq): Future[ParseHandler[Arrival]] =
```
Comment (amaalali) [inline app/services/ArrivalMovementMessageService.scala:42]: 👍

File: app/services/ArrivalMovementMessageService.scala [hunk 5]
```diff
@@ -80,10 +77,27 @@ class ArrivalMovementMessageService @Inject()(arrivalIdRepository: ArrivalIdRepo
       xmlMessage <- ReaderT[ParseHandler, NodeSeq, NodeSeq](nodeSeqToEither)
     } yield MovementMessageWithoutStatus(dateTime, messageType, xmlMessage, messageCorrelationId)
 
+  //TODO: Not sure if this is used anymore
   def makeMovementMessageWithStatus(messageCorrelationId: Int, messageType: MessageType): ReaderT[ParseHandler, NodeSeq, MovementMessageWithStatus] =
     for {
       _          <- correctRootNodeR(messageType)
       dateTime   <- dateTimeOfPrepR
       xmlMessage <- ReaderT[ParseHandler, NodeSeq, NodeSeq](nodeSeqToEither)
     } yield MovementMessageWithStatus(dateTime, messageType, xmlMessage, SubmissionPending, messageCorrelationId)
```
Comment (amaalali) [inline app/services/ArrivalMovementMessageService.scala:86]: If it looks like dead code we could get rid of it

**Review by amaalali — APPROVED** [review]

File: app/services/ArrivalMovementMessageService.scala (hunk 5 above)
Comment (chris-ianson) [reply app/services/ArrivalMovementMessageService.scala]: Removed and renamed the methods to make more sense

**Review by amaalali — APPROVED** [review]

**Pull request merged**